src/System.IO/src/System/IO/BufferedStream.cs

```diff
@@ -0,0 +1,682 @@
+    /// alternate between the two of them on the same stream.
+    ///
+    /// Class Invariants:
+    /// The class has one buffer, shared for reading & writing.
+    /// It can only be used for one or the other at any point in time - not both.
+    /// The following should be true:
+    /// <![CDATA[
+    ///   * 0 <= _readPos <= _readLen < _bufferSize
+    ///   * 0 <= _writePos < _bufferSize
+    ///   * _readPos == _readLen && _readPos > 0 implies the read buffer is valid, but we're at the end of the buffer.
+    ///   * _readPos == _readLen == 0 means the read buffer contains garbage.
+    ///   * Either _writePos can be greater than 0, or _readLen & _readPos can be greater than zero,
+    ///     but neither can be greater than zero at the same time.
+    ///  ]]>
+    /// This class will never cache more bytes than the max specified buffer size.
+    /// However, it may use a temporary buffer of up to twice the size in order to combine several IO operations on
+    /// the underlying stream into a single operation. This is because we assume that memory copies are significantly
+    /// faster than IO operations on the underlying stream (if this was not true, using buffering is never appropriate).
+    /// The max size of this "shadow" buffer is limited as to not allocate it on the LOH.
+    /// Shadowing is always transient. Even when using this technique, this class still guarantees that the number of
+    /// bytes cached (not yet written to the target stream or not yet consumed by the user) is never larger than the 
+    /// actual specified buffer size.
+    /// </summary>
+    [ComVisible(true)]
+    public sealed partial class BufferedStream : Stream
+    {
+        private const Int32 _DefaultBufferSize = 4096;
+
+
+        private Stream _stream;                               // Underlying stream.  Close sets _stream to null.
+
+        private Byte[] _buffer;                               // Shared read/write buffer.  Alloc on first use.
+
+        private readonly Int32 _bufferSize;                   // Length of internal buffer (not counting the shadow buffer).
+
+        private Int32 _readPos;                               // Read pointer within shared buffer.
+        private Int32 _readLen;                               // Number of bytes read in buffer from _stream.
+        private Int32 _writePos;                              // Write pointer within shared buffer.
+
+        /// <summary><code>MaxShadowBufferSize</code> is chosed such that shadow buffers are not allocated on the Large Object Heap.
+        /// Currently, an object is allocated on the LOH if it is larger than 85000 bytes. See LARGE_OBJECT_SIZE in ndp\clr\src\vm\gc.h
+        /// We will go with exactly 80 KBytes, although this is somewhat arbitrary.</summary>
+        private const Int32 MaxShadowBufferSize = 81920;  // Make sure not to get to the Large Object Heap.
```

Comment:
Nit: Why is this called a "Shadow" buffer, when other references like "_buffer" and "_bufferSize" don't include that term?